bignum/src/commonMain/kotlin/com/ionspin/kotlin/bignum/decimal/BigDecimal.kt

```diff
@@ -1231,6 +1261,97 @@ class BigDecimal private constructor(
      */
     override fun signum(): Int = significand.signum()
 
+    /**
+     * The next group of functions are implementations of the NarrowingOperations interface
+     */
+    /**
+     * Convert the current value to  Int.
+     * @param exactRequired True causes an ArithmeticException to be thrown if there is any loss of
+     * precision during the conversion, either side of the decimal. False truncates any precision to
+     * the right of the decimal.
+     */
+    override fun intValue(exactRequired: Boolean): Int {
+        return significand.intValue(exactRequired)
```

Comment:
This won't be correct because it doesn't consider the exponent. Currently the BigDecimal is constructed like this:
`significand * exponent`. 
For example, if significand is 987654321 and exponent is 5, then resulting big decimal will be 9.87654321 * 10^5 = 987654.321

Each BigDecimal has also a precision member, which is just the number of digits in significand


You can see the current state if you run this code snippet:
```        
        val bigDecimal = BigDecimal.fromLongWithExponent(987654321, 5)
        println("Value: ${bigDecimal.toStringExpanded()}")
        val longValue = bigDecimal.longValue()
        println("Long value $longValue")
```

I think you can reuse the `BigDecimal.toBigInteger()`  because this will return a BigInteger that already has truncated the non-integer part or multiplied by a certain power and then we can narrow that BigInteger to long/int or whatever was requested. 

We would also need to check if the number that we are returning is exact as the original when `exactRequired=true`, as 123.4 will be returned as 123. You might check that by calling `.floor()` on the bigDecimal and checking if it's zero or not. Or there might be a more efficient way, I'm not completely sure about that part.

Reply:
Cool, fixing now. Also if it's OK I'd like to add these changes to the BigInteger narrowing functions using the unsigned types.  If I change this in longValue as an example:
```
   if (exactRequired && (this > Long.MAX_VALUE.toULong())) {
```
to this, it fixes an issue when this is negative:
```
        if (exactRequired && (this.abs() > Long.MAX_VALUE.toULong())) {
```

Reply:
Oops I think I clicked on the wrong thing and caused another build.  Anyway I think build failed cuz of my new tests failing. I'm gonna commit changes to address the issue above, and it also makes small changes to BigInteger.  In my environment the tests pass now. So won't be long and I'll post again, as soon as I figure out how to update a pull with a new commit :-)

Reply:
Question - on the unsigned narrows it needs to check for negative numbers and bail on those when exactRequired is true.  As part of making that change to both BigDecimal and BigInteger, would it be ok to add these two properties to the BigNumber interface? The java equivalents are handy...

```
    val isNegative get() = signum() < 0
    val isPositive get() = signum() > 0
```

Reply:
Let me know if that's ok, and I'll do a commit to the pull that hopefully gets the tests working again

Reply:
> Also if it's OK I'd like to add these changes to the BigInteger narrowing functions using the unsigned types

Absolutely fine with adding that here, but I'm not sure if we can use abs(), because absolute value of BigInteger with value of Long.MIN_VALUE is 9223372036854775808 and we are comparing against Long.MAX_VALUE, which is one less at 9223372036854775807, so it would fail in that one case. I think we need to compare against both MIN_VALUE and MAX_VALUE

> Question - on the unsigned narrows it needs to check for negative numbers and bail on those when exactRequired is true. 

Once again, you are completely right we need that check if the BigInt is negative if exact is required, I'd just suggest either reusing the existing `sign` member of BigInteger (it's an enum with POSITIVE, NEGATIVE, ZERO value) or add the helpers as (there we save one call to signum() which does the same thing)
```
   val isNegative get() = sign == Sign.NEGATIVE
   val isPositive get() = sign == Sign.POSITIVE
```
And don't worry about restarting the builds, Travis CI offers them for free for open source projects, and the build for this library is not that long in any case.

Reply:
Ok, I got rid of the UInt usage in byteValue, shortValue, intValue and longValue, and inserted range checks. Gonna do a commit and figure out how to get the pull request updated.

Reply:
Awesome! The pull request will update automatically when you push to your original branch, and it should also run the Travis build on the latest commit at the same time